pkg/algorithm/nodeselect/v1alpha2/select_node.go

```diff
@@ -0,0 +1,206 @@
+		}
+	}
+	return candidateNodesMap, nil
+}
+
+// GetUsedNodeMap returns a map of node for which pool has already been created.
+func (ac *Config) GetUsedNodeMap() (map[string]bool, error) {
+	usedNodeMap := make(map[string]bool)
+	cspList, err := csp.
+		NewKubeClient().
+		WithNamespace(ac.Namespace).
+		List(
+			metav1.
+				ListOptions{LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + ac.CSPC.Name},
+		)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list already created csp(s)")
+	}
+	for _, cspObj := range cspList.Items {
+		usedNodeMap[cspObj.Labels[string(apis.HostNameCPK)]] = true
+	}
+	return usedNodeMap, nil
+}
+
+// SelectNode selects a node and returns the pool spec from the cspc for pool provisioning.
+func (ac *Config) SelectNode() (*apis.PoolSpec, error) {
+	candidateNodes, err := ac.GetCandidateNodeMap()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get pool spec for pool creation")
+	}
+	for _, pool := range ac.CSPC.Spec.Pools {
+		pool := pool
+		nodeName := pool.NodeSelector[HostName]
+		if candidateNodes[nodeName] {
+			if ValidatePoolSpec(&pool) {
+				return &pool, nil
+			}
+		}
+	}
+	return nil, errors.New("no node qualified for pool creation")
+}
+
+// GetBDListForNode returns a list of BD from the pool spec.
+// TODO : Move it to CStorPoolCluster packgage
+func (ac *Config) GetBDListForNode(pool *apis.PoolSpec) []string {
+	var BDList []string
+	for _, group := range pool.RaidGroups {
+		for _, bd := range group.BlockDevices {
+			BDList = append(BDList, bd.BlockDeviceName)
+		}
+	}
+	return BDList
+}
+
+// ClaimBDsForNode claims a given BlockDevice for node
+// If the block device(s) is/are already claimed for any other CSPC it returns error.
+// If the block device(s) is/are already calimed for the same CSPC -- it is left as it is and can be used for
+// pool provisioning.
+// If the block device(s) is/are unclaimed, then those are claimed.
+func (ac *Config) ClaimBDsForNode(BD []string) error {
+	for _, bd := range BD {
+		IsBDClaimed, err := ac.IsBDClaimed(bd)
+		if err != nil {
+			return errors.Wrapf(err, "error in getting details for BD {%s} whether it is claimed", bd)
+		}
+		if IsBDClaimed {
+			IsClaimedBDUsable, err := ac.IsClaimedBDUsable(bd)
+			if err != nil {
+				return errors.Wrapf(err, "error in getting details for BD {%s} for usability", bd)
+			}
+			if !IsClaimedBDUsable {
+				return errors.Errorf("BD {%s} already in use", bd)
+			}
+		}
+	}
+	return ac.ClaimBDList(BD)
+}
+
+// ClaimBDList claims a list of  BlockDevices
+// If the block device is already claimed -- no action is performed i.e. it remains claimed
+func (ac *Config) ClaimBDList(BDList []string) error {
+	if len(BDList) == 0 {
+		return errors.New("No block devices to claim")
+	}
+	for _, bd := range BDList {
+		IsBDClaimed, err := ac.IsBDClaimed(bd)
+		if err != nil {
+			return errors.Wrapf(err, "error in getting details for BD {%s} whether it is claimed", bd)
+		}
+		if IsBDClaimed {
```

Comment:
can we check and claim then and there? i.e., merging this code with previous function